pykob/morse.py

```diff
@@ -212,41 +248,58 @@ def flush(self):
         if self.flusher:
             self.flusher.cancel()
             self.flusher = None
-        self.decodeChar(MAXINT)
-        self.decodeChar(MAXINT)  # this is intentional, although maybe not needed
+        if self.mark > 0 or self.latched:
+            spacing = self.spaceBuf[self.nChars]
+            if self.mark > MINDASHLEN * self.truDot:
+                self.codeBuf[self.nChars] += '-'  # dash
+            elif self.mark > 2:
+                self.codeBuf[self.nChars] += '.'  # dot
+            self.markBuf[self.nChars] = self.mark
+            self.mark = 0
+            self.space = 1  # to prevent circuit opening mistakenly decoding as 'E'
+            self.decodeChar(MAXINT)
+            self.decodeChar(MAXINT)  # a second time, to flush both characters
+            self.codeBuf = ['', '']
+            self.spaceBuf = [0, 0]
+            self.markBuf = [0, 0]
+            self.nChars = 0
+            if self.latched:
+                self.callback('_', float(spacing) / (3 * self.truDot) - 1)
 
     def decodeChar(self, nextSpace):
-        self.nChars += 1
-        sp1 = self.spaceBuf[0]
-        sp2 = self.spaceBuf[1]
-        sp3 = nextSpace
-        code = ''
-        s = ''
+        self.nChars += 1  # number of complete characters in buffer (1 or 2)
+        sp1 = self.spaceBuf[0]  # space before 1st character
+        sp2 = self.spaceBuf[1]  # space before 2nd character
+        sp3 = nextSpace  # space before next character
+        code = ''  # the dots and dashes
+        s = ''  # the decoded character or pair of characters
         if self.nChars == 2 and sp2 < MAXMORSESPACE * self.dotLen and \
-                MORSERATIO * sp1 > sp2 and sp2 < MORSERATIO * sp3:
-            code = self.codeBuf[0] + ' ' + self.codeBuf[1]
+                MORSERATIO * sp1 > sp2 and sp2 < MORSERATIO * sp3:  # could be two halves of a spaced character
+            code = self.codeBuf[0] + ' ' + self.codeBuf[1]  # try combining the two halves
             s = self.lookupChar(code)
-            if s != '' and s != '&':
+            if s != '' and s != '&':  # yes, it's a spaced character, clear the whole buffer
```

Comment:
I really like these comments (this, plus the ones below)! They really help when working with the code.

Reply:
The comments are the bare minimum that I needed to keep things straight in my own mind. If they're helpful to someone else, so much the better!